conbench/api/results.py

```diff
@@ -182,14 +186,86 @@ def post(self) -> None:
         tags:
           - Benchmarks
         """
-        # Here it sould be easy to make `data` have a precise type (that mypy
+        # Here it should be easy to make `data` have a precise type (that mypy
         # can use) based on the schema that we validate against.
         data = self.validate_benchmark(self.schema.create)
+
+        # Note(JP): this next block inspects and mutates `tags`, with the goal
+        # that all keys are non-empty strings, and all values are non-empty
+        # strings. See
+        # https://github.com/conbench/conbench/pull/948#discussion_r1149090197
+        # for background. Summary of current desired behavior: primitive value
+        # types are accepted (string, boolean, float, int; non-string values
+        # are converted to string before DB insertion). Other value types
+        # (array -> list, object -> dict) lead to request rejection.
+        tags = data["tags"]
+        # Iterate over a copy of key/value pairs.
+        for key, value in list(tags.items()):
+            # In JSON, a key is always of type string. We rely on this, codify
+            # this invariant.
+            assert isinstance(key, str)
+
+            # An empty string is a valid JSON key. Do not allow this.
+            if len(key) == 0:
+                return resp400("tags: zero-length string as key is not allowed")
+
+            # For now, be liberal in what we accept. Do not consider empty
+            # string or None values for the case permutation (do not store
+            # those in the DB, drop these key/value pairs). This is documented
+            # in the API spec. Maybe in the future we want to reject such
+            # requests with a Bad Request response.
+            if value == "" or value is None:
+                log.warning("drop tag key/value pair: `%s`, `%s`", key, value)
+                # Remove current key/value pair, proceed with next key. This
+                # mutates the dictionary `data["tags"]`; for keeping this a
+                # sane operation the loop iterates over a copy of key/value
+                # pairs.
+                del tags[key]
+                continue
+
+            # Note(JP): this code path should go away after we adjust our
+            # client tooling to not send numeric values anymore.
+            if isinstance(value, (int, float, bool)):
+                # I think we first want to adjust some client tooling before
+                # enabling this log line:
+                # log.warning("stringify case parameter value: `%s`, `%s`", key, value)
+                # Replace value, proceed with next key.
+                tags[key] = str(value)
```

Comment:
@boshek A little heads-up! I'm not sure offhand if any of the reporting relies on non-string tag values, but if so you may need to cast back